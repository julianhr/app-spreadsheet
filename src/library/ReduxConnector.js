import appStore from '~/reducers/'


class ReduxConnector {
  constructor(store=appStore) {
    this.store = store
    this.funcs = {}
  }

  get fn() {
    return this.funcs
  }

  getState() {
    return this.store.getState()
  }

  registerFunctions(funcs) {
    for (let [name, fn] of Object.entries(funcs)) {
      this.funcs[name] = async function(...args) {
        await this.store.dispatch( fn(...args) )
      }.bind(this)
    }
  }
}

export default ReduxConnector
