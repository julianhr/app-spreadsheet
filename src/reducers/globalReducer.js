import { createReducer } from 'redux-starter-kit'

export const INITIAL_STATE = {
  activeCell: null,
  rows: 14,
  columns: 6,
}

export default createReducer(INITIAL_STATE, {
  'SET_ACTIVE_CELL': (state, { payload }) => {
    state.activeCell = payload
  },
})
