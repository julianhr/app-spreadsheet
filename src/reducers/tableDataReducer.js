import { createReducer } from 'redux-starter-kit'


const INITIAL_STATE = {}

export default createReducer(INITIAL_STATE, {
  'SET_CELL_DATA': (state, { payload }) => {
    const { location, data } = payload
    state[location] = data
  },

  'DELETE_CELL_DATA': (state, { payload }) => {
    delete state[payload]
  }
})
