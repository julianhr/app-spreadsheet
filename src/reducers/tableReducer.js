import { createReducer } from 'redux-starter-kit'


const INITIAL_STATE = {}

export default createReducer(INITIAL_STATE, {
  'SET_CELL_VALUE': (state, { payload }) => {
    const { location, text, formula } = payload
    state[location] = { text, formula }
  },

  'DELETE_CELL_VALUE': (state, { payload }) => {
    delete state[payload]
  }
})
