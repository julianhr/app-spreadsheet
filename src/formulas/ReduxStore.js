import ReduxConnector from '~/library/ReduxConnector'
import { clearCellData, setCellData } from '~/actions/tableDataActions'


const reduxConnector = new ReduxConnector()

const dispatchFunctions = {
  clearCellData,
  setCellData,
}

reduxConnector.registerFunctions(dispatchFunctions)

class ReduxStore {
  constructor(rc=reduxConnector) {
    const { rows, columns } = rc.getState().global
    this.rows = rows
    this.columns = columns
    this.reduxConnector = rc
  }

  get tableData() {
    return this.reduxConnector.getState().tableData
  }

  get locations() {
    return Object.keys(this.tableData)
  }

  getCellResult(location) {
    const data = this.tableData[location]
    return data && data.result
  }

  setCellData(location, entered, result) {
    this.reduxConnector.fn.setCellData(location, entered, result)
  }

  clearCellData(location) {
    this.reduxConnector.fn.clearCellData(location)
  }
}

export default ReduxStore
export { dispatchFunctions }
