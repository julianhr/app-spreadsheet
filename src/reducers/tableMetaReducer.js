import { createReducer } from 'redux-starter-kit'
import { DEFAULT_COL_WIDTH } from '~/library/constants'


const INITIAL_STATE = {
  colWidths: {},
}

export default createReducer(INITIAL_STATE, {
  'SET_COL_WIDTH': (state, { payload }) => {
    const { colLabel, width } = payload
    const { colWidths } = state
    
    if (width === DEFAULT_COL_WIDTH) {
      delete state.colWidths[colLabel]
    } else {
      state.colWidths = { ...colWidths, [colLabel]: width }
    }
  },
})
