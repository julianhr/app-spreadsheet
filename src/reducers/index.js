import { configureStore } from 'redux-starter-kit'

import globalReducer from './globalReducer'
import tableReducer from './tableReducer'


const appStoreFn = () => (
  configureStore({
    reducer: {
      global: globalReducer,
      table: tableReducer,
    }
  })
)


export default appStoreFn
