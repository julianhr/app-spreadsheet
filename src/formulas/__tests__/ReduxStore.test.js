import { appStoreGen } from '~/reducers/'
import ReduxStore, { dispatchFunctions } from '../ReduxStore'
import ReduxConnector from '~/library/ReduxConnector'


describe('ReduxStore', () => {
  let rs

  beforeEach(() => {
    const store = appStoreGen()
    const reduxConnector = new ReduxConnector(store)
    reduxConnector.registerFunctions(dispatchFunctions)
    rs = new ReduxStore(reduxConnector)
  })

  test('instantiation', () => {
    expect(rs).toMatchSnapshot()
  })

  test('get #tableData', () => {
    rs.setCellData('A-1', '7', 7)
    rs.setCellData('B-2', '=A1+3', 10)
    expect(rs.tableData).toMatchSnapshot()
  })

  test('get #locations', () => {
    rs.setCellData('A-1', '7', 7)
    rs.setCellData('B-2', '=A1+3', 10)
    expect(rs.locations).toEqual(['A-1', 'B-2'])
  })

  test('#getCellResult', () => {
    rs.setCellData('B-2', '=A1+3', 3)
    expect(rs.getCellResult('B-2')).toBe(3)
  })

  test('#setCellData', () => {
    rs.setCellData('B-2', '=A1+3', 3)
    expect(rs.tableData).toMatchSnapshot()

    rs.setCellData('B-2', '=c5/7', 0)
    expect(rs.tableData).toMatchSnapshot()
  })

  test('#clearCellData', () => {
    rs.setCellData('B-2', '=A1+3', 3)
    expect(rs.getCellResult('B-2')).toBe(3)

    rs.clearCellData('B-2')
    expect(rs.getCellResult('B-2')).toBeUndefined()
  })
})
