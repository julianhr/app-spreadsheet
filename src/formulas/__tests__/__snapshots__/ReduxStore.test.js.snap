// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReduxStore #setCellData 1`] = `
Object {
  "B-2": Object {
    "entered": "=A1+3",
    "result": 3,
  },
}
`;

exports[`ReduxStore #setCellData 2`] = `
Object {
  "B-2": Object {
    "entered": "=c5/7",
    "result": 0,
  },
}
`;

exports[`ReduxStore get #tableData 1`] = `
Object {
  "A-1": Object {
    "entered": "7",
    "result": 7,
  },
  "B-2": Object {
    "entered": "=A1+3",
    "result": 10,
  },
}
`;

exports[`ReduxStore instantiation 1`] = `
ReduxStore {
  "columns": 10,
  "reduxConnector": ReduxConnector {
    "funcs": Object {
      "clearCellData": [Function],
      "setCellData": [Function],
    },
    "store": Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    },
  },
  "rows": 30,
}
`;
